File: day18_python/soln.py
import functools
from math import floor, ceil


def explode(s):
    left = []
    brace_cnt = 0
    for c in s:
        match c:
            case '[':
                brace_cnt += 1
            case ']':
                brace_cnt -= 1
                if brace_cnt < 0:
                    raise 'unexpected ]'
            case n if c.isdigit() and brace_cnt > 4:
                next = len(left) + 1
                if s[next] == ',' and s[next + 1].isdigit():
                    # explode left
                    l = n
                    r = s[next + 1]
                    for i, c in enumerate(reversed(left)):
                        i = len(left) - 1 - i
                        if c.isdigit():
                            left[i] = str(int(c) + int(l))
                            break
                    # explode right
                    rest = list(s[len(left) + 4:])
                    for i, c in enumerate(rest):
                        if c.isdigit():
                            rest[i] = str(int(c) + int(r))
                            break
                    return ''.join(left[:-1] + ['0'] + rest)
            case _ if c.isdigit():
                pass
            case ',':
                pass
            case otherwise:
                raise f'unexpected: ${otherwise}'
        left.append(c)

    # no explosion happened
    return None


def split(orig_s):
    s = list(orig_s)
    for i, c in enumerate(s):
        if c.isdigit():
            length = 1
            while orig_s[i + length].isdigit():
                length += 1
            if length > 1:
                num = int(orig_s[i:i + length]) / 2
                l = floor(num)
                r = ceil(num)
                s[i] = f'[{l},{r}]'
                s[i + 1] = ''
                return ''.join(s)
    return None


def reduce(s):
    while True:
        res = explode(s)
        if res:
            s = res
            print("explode", s)
            continue

        res = split(s)
        if res:
            s = res
            print("split", s)
            continue
        break
    return s


def add(l, r):
    return reduce('[' + l + ',' + r + ']')


print(functools.reduce(add, open('sample').read().splitlines()))

File: day18_ruby/soln.rb
def tokenize(s)
  tokens = []

  i = 0
  while i < s.size
    if s[i].match?(/[0-9]/)
      num = s[i].to_i
      i += 1
      while s[i].match?(/[0-9]/)
        num = num * 10 + s[i].to_i
        i += 1
      end
      tokens << num
    else
      tokens << s[i]
      i += 1
    end
  end

  tokens
end

def explode(tokens)
  bracket_stack = 0
  (0...tokens.size).each do |i|
    c = tokens[i]
    if c == '['
      bracket_stack += 1
    elsif c == ']'
      bracket_stack -= 1
    elsif c.is_a?(Integer) && bracket_stack > 4
      # do explosion
      l = c
      r = tokens[i + 2]

      # explode left
      ((i - 1)..0).step(-1).each do |i|
        if tokens[i].is_a?(Integer)
          tokens[i] += l
          break
        end
      end

      # explode right
      ((i + 3)...tokens.size).each do |i|
        if tokens[i].is_a?(Integer)
          tokens[i] += r
          break
        end
      end

      return tokens[0..(i - 2)] + [0] + tokens[(i + 4)..]
    end
  end

  nil
end

def split(tokens)
  (0...tokens.size).each do |i|
    if tokens[i].is_a?(Integer) && tokens[i] >= 10
      l = (tokens[i] / 2.0).floor
      r = (tokens[i] / 2.0).ceil
      return tokens[0...i] + ['[', l, ',', r, ']'] + tokens[(i + 1)..]
    end
  end
  nil
end

def reduce(tokens)
  loop do
    # first try explode
    exploded = explode(tokens)
    unless exploded.nil?
      tokens = exploded
      next
    end

    # then try split
    split = split(tokens)
    unless split.nil?
      tokens = split
      next
    end

    break
  end
  tokens
end

def add(l, r)
  reduce(['['] + l + [','] + r + [']'])
end

def magnitude(tokens)
  magnitude_arr(eval(tokens.map(&:to_s).join))
end

def magnitude_arr(arr)
  return arr if arr.is_a?(Integer)
  lhs, rhs = arr
  (3 * magnitude_arr(lhs) + 2 * magnitude_arr(rhs))
end

inputs = File.read("input").split("\n")
final = tokenize(inputs.first)
inputs[1..].each do |input|
  final = add(final, tokenize(input))
end
p final.map(&:to_s).join
p magnitude(final)
